File: src/llm_router/services/rate_limit.py
from __future__ import annotations

import asyncio
import time
from typing import Dict, Iterable, Optional

from ..db.models import RateLimit as RateLimitModel
from ..schemas import RateLimitConfig


class TokenBucket:
    def __init__(self, config: RateLimitConfig) -> None:
        self.max_requests = config.max_requests
        self.per_seconds = config.per_seconds
        self.burst_size = config.burst_size or config.max_requests
        self._refill_rate = self.max_requests / self.per_seconds
        self._tokens = float(self.burst_size)
        self._last_refill = time.monotonic()
        self._lock = asyncio.Lock()

    def _refill(self) -> None:
        now = time.monotonic()
        elapsed = now - self._last_refill
        self._last_refill = now
        refill = elapsed * self._refill_rate
        if refill > 0:
            self._tokens = min(self.burst_size, self._tokens + refill)

    async def acquire(self, tokens: int = 1) -> None:
        if tokens <= 0:
            return

        while True:
            async with self._lock:
                self._refill()
                if self._tokens >= tokens:
                    self._tokens -= tokens
                    return
                deficit = tokens - self._tokens
                wait_time = deficit / self._refill_rate if self._refill_rate > 0 else self.per_seconds
            await asyncio.sleep(wait_time)


class RateLimiterManager:
    def __init__(self) -> None:
        self._buckets: Dict[int, TokenBucket] = {}
        self._lock = asyncio.Lock()

    def upsert(self, model_id: int, config: RateLimitConfig) -> None:
        self._buckets[model_id] = TokenBucket(config)

    def remove(self, model_id: int) -> None:
        self._buckets.pop(model_id, None)

    def load_from_records(self, records: Iterable[RateLimitModel]) -> None:
        for record in records:
            config = RateLimitConfig(
                max_requests=record.max_requests,
                per_seconds=record.per_seconds,
                burst_size=record.burst_size,
                notes=record.notes,
                config=record.config,
            )
            self.upsert(record.model_id, config)

    async def acquire(self, model_id: int, tokens: int = 1) -> None:
        bucket = self._buckets.get(model_id)
        if not bucket:
            return
        await bucket.acquire(tokens)

    def get_bucket(self, model_id: int) -> Optional[TokenBucket]:
        return self._buckets.get(model_id)

File: src/services/rate_limit.go
package services

import (
	"sync"
	"time"
)

// RateLimitConfig mirrors python service-level token bucket config.
type RateLimitConfig struct {
	MaxRequests int64
	PerSeconds  int64
	BurstSize   int64
}

type tokenBucket struct {
	maxRequests float64
	perSeconds  float64
	burstSize   float64
	refillRate  float64
	tokens      float64
	lastRefill  time.Time
	mu          sync.Mutex
}

func newTokenBucket(cfg RateLimitConfig) *tokenBucket {
	burst := cfg.BurstSize
	if burst <= 0 {
		burst = cfg.MaxRequests
	}
	if cfg.MaxRequests <= 0 {
		cfg.MaxRequests = 1
	}
	if cfg.PerSeconds <= 0 {
		cfg.PerSeconds = 1
	}
	return &tokenBucket{
		maxRequests: float64(cfg.MaxRequests),
		perSeconds:  float64(cfg.PerSeconds),
		burstSize:   float64(burst),
		refillRate:  float64(cfg.MaxRequests) / float64(cfg.PerSeconds),
		tokens:      float64(burst),
		lastRefill:  time.Now(),
	}
}

func (b *tokenBucket) refill(now time.Time) {
	elapsed := now.Sub(b.lastRefill).Seconds()
	if elapsed <= 0 {
		return
	}
	b.lastRefill = now
	b.tokens += elapsed * b.refillRate
	if b.tokens > b.burstSize {
		b.tokens = b.burstSize
	}
}

func (b *tokenBucket) acquire(tokens int64) {
	if tokens <= 0 {
		return
	}
	need := float64(tokens)
	for {
		b.mu.Lock()
		b.refill(time.Now())
		if b.tokens >= need {
			b.tokens -= need
			b.mu.Unlock()
			return
		}
		deficit := need - b.tokens
		waitSeconds := deficit / b.refillRate
		b.mu.Unlock()
		if waitSeconds <= 0 {
			waitSeconds = 0.001
		}
		time.Sleep(time.Duration(waitSeconds * float64(time.Second)))
	}
}

// RateLimiterManager applies per-model token bucket controls.
type RateLimiterManager struct {
	mu      sync.RWMutex
	buckets map[int64]*tokenBucket
}

func NewRateLimiterManager() *RateLimiterManager {
	return &RateLimiterManager{buckets: map[int64]*tokenBucket{}}
}

func (m *RateLimiterManager) Upsert(modelID int64, cfg RateLimitConfig) {
	m.mu.Lock()
	m.buckets[modelID] = newTokenBucket(cfg)
	m.mu.Unlock()
}

func (m *RateLimiterManager) Remove(modelID int64) {
	m.mu.Lock()
	delete(m.buckets, modelID)
	m.mu.Unlock()
}

func (m *RateLimiterManager) Acquire(modelID int64, tokens int64) {
	m.mu.RLock()
	bucket := m.buckets[modelID]
	m.mu.RUnlock()
	if bucket == nil {
		return
	}
	bucket.acquire(tokens)
}
